### PR TITLE
generate checksum on the fly

### PR DESCRIPTION
**Pull request by dassio**
this is based on #19351 
>Keeping hashes for files stored on the server is a much more complex problem then just returning a hash from uploaded content.
In order to keep hashes for stored files, we need to be 100% sure that we update (or invalidate) the hash for every change made to the file on the server

 checksum generation for local file on the fly seems to be a good **central control point**.

>Adding this is tricky. Just imagine on a big instance and everybody starts to request the hash of several GB files. This could easily kill the server.

on an linode server with 1vCPU and 2G  RAM, this seems to be ok with me, I tried the following :
*  folder with **101**   files and the total size of **20G**,  propfind request  return time: **0:02:01**
*  folder with **1001**  files and total **20G**, propfind request return time: **0:01:36**

of course, the CPU is 100% during this time 

but this is a functionality issue, we can fix the performance issue later, generating checksum will always use lots of CPU . we can not avoid this, some methods that can minimize the performance impact : 
* cache the checksum result 
* check the etag first, only when etag is different check the checksum

### CODE REVIEW

File: apps/dav/lib/Connector/Sabre/FilesPlugin.php [hunk 1]
```diff
@@ -382,6 +384,7 @@ public function handleGetProperties(PropFind $propFind, \Sabre\DAV\INode $node)
 
 		if ($node instanceof \OCA\DAV\Connector\Sabre\File) {
 			$propFind->handle(self::DOWNLOADURL_PROPERTYNAME, function () use ($node) {
+				/** @var $node \OCA\DAV\Connector\Sabre\File */
```
Comment (kesselb) [inline apps/dav/lib/Connector/Sabre/FilesPlugin.php:387]: Redundant. Your ide should be able to detect the correct type from the instanxeof check.

File: apps/dav/lib/Connector/Sabre/FilesPlugin.php [hunk 2]
```diff
@@ -398,7 +401,12 @@ public function handleGetProperties(PropFind $propFind, \Sabre\DAV\INode $node)
 			$propFind->handle(self::CHECKSUMS_PROPERTYNAME, function () use ($node) {
 				$checksum = $node->getChecksum();
 				if ($checksum === null || $checksum === '') {
-					return null;
+          if($node->getFileInfo()->getMountPoint() instanceof ExternalMountPoint) {
```
Comment (kesselb) [inline apps/dav/lib/Connector/Sabre/FilesPlugin.php:404]: Space before bracket

File: apps/dav/lib/Connector/Sabre/FilesPlugin.php [hunk 3]
```diff
@@ -398,7 +401,12 @@ public function handleGetProperties(PropFind $propFind, \Sabre\DAV\INode $node)
...
+            return null;
+          }else{
```
Comment (kesselb) [inline apps/dav/lib/Connector/Sabre/FilesPlugin.php:406]: } else {

File: apps/dav/lib/Connector/Sabre/FilesPlugin.php [hunk 4]
```diff
@@ -514,5 +522,25 @@ public function sendFileIdHeader($filePath, \Sabre\DAV\INode $node = null) {
 				$this->server->httpResponse->setHeader('OC-FileId', $fileId);
 			}
 		}
-	}
+  }
+
+  private function generateLoclFileChecksum(\Sabre\DAV\INode $node): String {
```
Comment (kesselb) [inline apps/dav/lib/Connector/Sabre/FilesPlugin.php:527]: The return type should be `string` (lowercase)

Comment (kesselb) [inline apps/dav/lib/Connector/Sabre/FilesPlugin.php:527]: Locl to Local

File: apps/dav/lib/Connector/Sabre/FilesPlugin.php [hunk 5]
```diff
@@ -514,5 +522,25 @@ public function sendFileIdHeader($filePath, \Sabre\DAV\INode $node = null) {
...
+    $fileStorage = $node->getFileInfo()->getStorage(); 
+    $source = $fileStorage->fopen($node->getFileInfo()->getInternalPath(),'r'); 
+    fseek($source, 0);
+
+    $obtainedHash = null;
```
Comment (kesselb) [inline apps/dav/lib/Connector/Sabre/FilesPlugin.php:532]: Should be '' or the return type ?string

**Review by kesselb — COMMENTED** [review]
Thanks for sending a patch 👍

File: apps/dav/lib/Connector/Sabre/FilesPlugin.php (hunk 2 above)
Comment (kesselb) [inline apps/dav/lib/Connector/Sabre/FilesPlugin.php:404]: It seems the import for ExternalMountPoint is missing.

File: apps/dav/lib/Connector/Sabre/FilesPlugin.php (hunk 1 above)
Comment (dassio) [reply apps/dav/lib/Connector/Sabre/FilesPlugin.php]: removed

File: apps/dav/lib/Connector/Sabre/FilesPlugin.php (hunk 2 above)
Comment (dassio) [reply apps/dav/lib/Connector/Sabre/FilesPlugin.php]: add space

File: apps/dav/lib/Connector/Sabre/FilesPlugin.php (hunk 3 above)
Comment (dassio) [reply apps/dav/lib/Connector/Sabre/FilesPlugin.php]: add space

File: apps/dav/lib/Connector/Sabre/FilesPlugin.php (hunk 4 above)
Comment (dassio) [reply apps/dav/lib/Connector/Sabre/FilesPlugin.php]: change to lowercase

Comment (dassio) [reply apps/dav/lib/Connector/Sabre/FilesPlugin.php]: change method name

File: apps/dav/lib/Connector/Sabre/FilesPlugin.php (hunk 5 above)
Comment (dassio) [reply apps/dav/lib/Connector/Sabre/FilesPlugin.php]: using default empty string

File: apps/dav/lib/Connector/Sabre/FilesPlugin.php (hunk 2 above)
Comment (dassio) [reply apps/dav/lib/Connector/Sabre/FilesPlugin.php]: add import

**Comment by Dragon988** [comment]
Why are we using sha1 instead of sha256 or crc32? Modern processors can hardware accelerate both.

**Review by SimplyCorbett — CHANGES_REQUESTED** [review]
$stream = HashWrapper::wrap($source, "sha1", $callback);

Should be

$stream = HashWrapper::wrap($source, "sha256", $callback);

**Comment by dassio** [comment]
> Why are we using SHA-1 instead of SHA-2 or crc32? Modern processors can hardware accelerate both.
> 
> Edit: Not an expert on these things but crc32 is good enough for error detection. It's useless for security but if we care about that then SHA-1 has been deprecated in a lot of places. Researchers have achieved a collision attack against SHA-1 successfully. We should be using SHA-2 if we care for security.
> 
> Edit2: Considering nextcloud is used in businesses including healthcare SHA-2 should be used by default.

I think checksum here is not more about security than  about change detection, but anyhow I tried  it on the same server with `sha256`, I am ok with changing to `sha256`

### Before 
folder with 1001 files and total 20G, propfind request return time: 0:01:36
### After 
```
curl  -k -u username:password 'https://server.name/remote.php/webdav/test ' -X PROPFIND --data '<?xml version="1.0" encoding="UTF-8"?><d:propfind xmlns:d="DAV:">
>    <d:prop xmlns:oc="http://owncloud.org/ns">
>      <d:getlastmodified/>
>      <d:getcontentlength/>
>      <checksums xmlns="http://owncloud.org/ns" />
>    </d:prop>
>  </d:propfind>' >> test1.xml 
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  388k    0  388k  100   242   1991      1  0:04:02  0:03:19  0:00:43 86268
```

**Comment by Dragon988** [comment]
> I think checksum here is not more about security than about change detection.

Well, when we think about it if checksums are usable on external storage then security would be a priority. I'm using healthcare as an example but they would go the extra length to verify files haven't been tampered with.

@dassio Will this work with nextclouds encryption enabled?

**Comment by dassio** [comment]
> @dassio Will this work with nextclouds encryption enabled?

yes, i tried , it works

File: apps/dav/lib/Connector/Sabre/FilesPlugin.php [hunk 6]
```diff
@@ -514,5 +521,25 @@ public function sendFileIdHeader($filePath, \Sabre\DAV\INode $node = null) {
...
+
+    $obtainedHash = "";
+    $callback = function($hash) use (&$obtainedHash) {
+        $obtainedHash = $hash;
+    };
+
+    $stream = HashWrapper::wrap($source, "sha256", $callback);
+
+    while(feof($stream) === false) {
+        fread($stream, 200);
```
Comment (kesselb) [inline apps/dav/lib/Connector/Sabre/FilesPlugin.php:539]: 200?

**Comment by dassio** [comment]
> **To fix: Lint / php-cs check (pull_request)**
> 
> Update the code to match the defined code style. The easiest way should be:
> 
> ```
> composer install
> composer run cs:fix
> ```
> 
> **To fix: Static code analysis / static-code-analysis (pull_request)**
> 
> Change
> 
> `private function generateLocalFileChecksum(\Sabre\DAV\INode $node): string {`
> 
> To
> 
> `private function generateLocalFileChecksum(\OCA\DAV\Connector\Sabre\File $node): string {`

there seems to be an problem instaling `psalm/phar (3.12.2)`, it is stuck at following line, i used a new version `3.14.0`

`  - Installing psalm/phar (3.12.2): Downloading (connecting...)`

now it should fix the lint issue.

File: apps/dav/lib/Connector/Sabre/FilesPlugin.php (hunk 6 above)
Comment (dassio) [reply apps/dav/lib/Connector/Sabre/FilesPlugin.php]: I copied this from somewhere else, I also tried with a different number, it seems to make no difference, so I stick with it 
 maybe you can propose some reasonable number each time it should read

**Comment by dassio** [comment]
@kesselb  @Dragon988 any update on this one, should we add the checksum in the server app or need implement in another app?

**Comment by kesselb** [comment]
Added labels for Nextcloud 21 and requested a review from team storage. 

Code style looks good, CI failures seems unrelated. 

I suspect this change will not work for people with much data and create a lot of Load/IO. Generating the checksum just to withdraw the result afterwards is bad. The upload case already reads the data (from stdin to store it afterwards) so calculating the hash is cheap. Here we are reading the data only to calculate the hash.

**Comment by icewind1991** [comment]
Even for local storage, calculating the hash simply doesn't scale if it has to be calculated beforehand, the only reason it can work on upload is because we can do it "inline" with reading the data from stdin and storing it on the storage

**Comment by skjnldsv** [comment]
Closing after @icewind1991's comment

**Pull request closed without merging**